NebulaPatcher/Patches/Dynamic/EnemyUnitComponent_Patch.cs

```diff
@@ -0,0 +1,400 @@
+                    double num17 = (double)hive.starData.systemRadius * 40000.0;
+                    if (num16 > num17 * num17 * 2.25)
+                    {
+                        if (@this.hatred.max.targetType == ETargetType.Craft)
+                        {
+                            @this.hatred.ClearMax();
+                        }
+                        return;
+                    }
+                    double num18 = x2 - x;
+                    double num19 = y2 - y;
+                    double num20 = z2 - z;
+                    double num21 = num18 * num18 + num19 * num19 + num20 * num20;
+                    if (num21 < (double)sqrSensorRange)
+                    {
+                        double num13 = (double)sqrSensorRange - Math.Sqrt(num21);
+                        if (num13 > 0.0)
+                        {
+                            @this.hatred.HateTarget(ETargetType.Craft, i, (int)num13, hatred_max_sense, EHatredOperation.Add);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static void PlayerSearchLogic(ref EnemyUnitComponent @this, ref EnemyData enemy,
+        EnemyDFHiveSystem hive, int hatred_max_sense, float sensorRange)
+    {
+        var starId = hive.starData.id;
+        // Only search if there are players in space
+        if (!Multiplayer.Session.Combat.ActivedStarsMechaInSpace.Contains(starId)) return;
+
+        SpaceSector sector = hive.sector;
+        var sqrSensorRange = sensorRange * sensorRange;
+        var reachRange = sensorRange * 1.45f + sector.skillSystem.playerEnergyShieldRadius;
+
+        // Convert enemy local pos to upos
+        sector.TransformFromAstro_ref(enemy.astroId, out var enemyUpos, ref enemy.pos);
+
+        // Find the closest alive player to the unit
+        var closestPlayerId = -1;
+        var cloestSqrDist = float.MaxValue;
+        var players = Multiplayer.Session.Combat.Players;
+        for (var i = 0; i < players.Length; i++)
+        {
+            ref var playerPosition = ref players[i];
+            if (playerPosition.starId != starId || !playerPosition.isAlive) continue;
+            var sqrDist = Vector3.SqrMagnitude(playerPosition.skillTargetU - enemyUpos);
+            if (sqrDist < cloestSqrDist)
+            {
+                closestPlayerId = playerPosition.id;
+                cloestSqrDist = sqrDist;
+            }
+        }
+
+        if (closestPlayerId == -1) // No alive player in the system
+        {
+            if (@this.hatred.max.targetType == ETargetType.Player)
+            {
+                @this.hatred.ClearMax();
+            }
+            return;
+        }
+
+        if (enemy.astroId == hive.hiveAstroId) // Enemy is in the hive
+        {
+            var sensorCoef = ((@this.hatred.max.targetType == ETargetType.Player) ? 1f : 0.64f);
+            if (cloestSqrDist > sqrSensorRange * sensorCoef)
+            {
+                return;
+            }
```

Comment:
Minor nitpick, could be simplified to:
```suggestion
            if (cloestSqrDist > sqrSensorRange * sensorCoef) return;
```